pyopenssl/19.0.0/pyopenssl-19.0.0.patch

```diff
@@ -0,0 +1,211 @@
+diff --git a/src/OpenSSL/SSL.py b/src/OpenSSL/SSL.py
+index 0687fc3..78e9c58 100644
+--- a/src/OpenSSL/SSL.py
++++ b/src/OpenSSL/SSL.py
+@@ -22,6 +22,8 @@ from OpenSSL._util import (
+     path_string as _path_string,
+     text_to_bytes_and_warn as _text_to_bytes_and_warn,
+     no_zero_allocator as _no_zero_allocator,
++    wolfssl as _wolfssl,
++    wolfssl_debug as _wolfssl_debug
+ )
+ 
+ from OpenSSL.crypto import (
+@@ -120,10 +122,12 @@ except NameError:
+ 
+ OPENSSL_VERSION_NUMBER = _lib.OPENSSL_VERSION_NUMBER
+ SSLEAY_VERSION = _lib.SSLEAY_VERSION
+-SSLEAY_CFLAGS = _lib.SSLEAY_CFLAGS
+-SSLEAY_PLATFORM = _lib.SSLEAY_PLATFORM
+-SSLEAY_DIR = _lib.SSLEAY_DIR
+-SSLEAY_BUILT_ON = _lib.SSLEAY_BUILT_ON
++
++if not _wolfssl:
++    SSLEAY_CFLAGS = _lib.SSLEAY_CFLAGS
++    SSLEAY_PLATFORM = _lib.SSLEAY_PLATFORM
++    SSLEAY_DIR = _lib.SSLEAY_DIR
++    SSLEAY_BUILT_ON = _lib.SSLEAY_BUILT_ON
+ 
+ SENT_SHUTDOWN = _lib.SSL_SENT_SHUTDOWN
+ RECEIVED_SHUTDOWN = _lib.SSL_RECEIVED_SHUTDOWN
+@@ -191,7 +195,7 @@ SESS_CACHE_NO_INTERNAL = _lib.SSL_SESS_CACHE_NO_INTERNAL
+ SSL_ST_CONNECT = _lib.SSL_ST_CONNECT
+ SSL_ST_ACCEPT = _lib.SSL_ST_ACCEPT
+ SSL_ST_MASK = _lib.SSL_ST_MASK
+-if _lib.Cryptography_HAS_SSL_ST:
++if not _wolfssl and _lib.Cryptography_HAS_SSL_ST:
+     SSL_ST_INIT = _lib.SSL_ST_INIT
+     SSL_ST_BEFORE = _lib.SSL_ST_BEFORE
+     SSL_ST_OK = _lib.SSL_ST_OK
+@@ -306,8 +310,22 @@ class _VerifyHelper(_CallbackExceptionHelper):
+         @wraps(callback)
+         def wrapper(ok, store_ctx):
+             x509 = _lib.X509_STORE_CTX_get_current_cert(store_ctx)
+-            _lib.X509_up_ref(x509)
+-            cert = X509._from_raw_x509_ptr(x509)
++            if _wolfssl:
++                # _from_raw_x509_ptr has code to call X509_free on the X509
++                # object when it's time to do garbage collection. With wolfSSL,
++                # when that X509_free call happens, it's possible that the X509
++                # object has already been freed by FreeX509 (an internal
++                # function). That function doesn't care what the reference count
++                # is and does the free unconditionally. This causes problems
++                # when the garbage collector comes along and tries to free the
++                # already freed X509. The solution for the wolfSSL case is to
++                # duplicate the object rather than fiddling with the ref count.
++                x509_copy = _lib.X509_dup(x509)
++                cert = X509._from_raw_x509_ptr(x509_copy)
```

Comment:
All calls to `FreeX509` should be changed to `wolfSSL_X509_free`. Leaving it like this is just going to cause problems down the line. I've found the following calls. Please change all calls to `wolfSSL_X509_free`. Then this section should not need a wolfSSL exception.
```
$ grep -rP '[^_]FreeX509\(' --include=*.[ch]
wolfssl/internal.h:    WOLFSSL_LOCAL void FreeX509(WOLFSSL_X509*);
src/ssl.c:        FreeX509(&ssl->peerCert);
src/ssl.c:static void ExternalFreeX509(WOLFSSL_X509* x509)
src/ssl.c:                FreeX509(x509);
src/ssl.c:    ExternalFreeX509(x509);
src/internal.c:void FreeX509(WOLFSSL_X509* x509)
src/internal.c:    FreeX509(&ssl->peerCert);
src/internal.c:                    FreeX509(x509);
src/internal.c:            FreeX509(x509);
src/internal.c:                            FreeX509(&ssl->peerCert);
src/internal.c:                            FreeX509(&ssl->peerCert);
```

Reply:
Obviously only `ExternalFreeX509` should be calling `FreeX509` (which means that `FreeX509` could be a static function in `ssl.c`.

Reply:
Makes sense to me. I'll create a wolfSSL PR today or Monday doing that (I'm off tomorrow).